tests/simulator.py

```diff
@@ -0,0 +1,69 @@
+"""Simulator(s) that can be used to create MSAL instance
+whose token cache is in a certain state, and remains unchanged.
+This generic simulator(s) becomes the test subject for different benchmark tools.
+
+For example, you can install pyperf and then run:
+
+    pyperf timeit -s "from tests.simulator import ClientCredentialGrantSimulator as T; t=T(tokens_per_tenant=1, cache_hit=True)" "t.run()"
+"""
+import json
+import logging
+import random
+from unittest.mock import patch
+
+import msal
+from tests.http_client import MinimalResponse
+
+
+logger = logging.getLogger(__name__)
+
+
+def _count_access_tokens(app):
+    return len(app.token_cache._cache[app.token_cache.CredentialType.ACCESS_TOKEN])
+
+
+class ClientCredentialGrantSimulator(object):
+
+    def __init__(self, number_of_tenants=1, tokens_per_tenant=1, cache_hit=False):
+        logger.info(
+            "number_of_tenants=%d, tokens_per_tenant=%d, cache_hit=%s",
+            number_of_tenants, tokens_per_tenant, cache_hit)
+        with patch.object(msal.authority, "tenant_discovery", return_value={
+            "authorization_endpoint": "https://contoso.com/placeholder",
+            "token_endpoint": "https://contoso.com/placeholder",
+        }) as _:  # Otherwise it would fail on OIDC discovery
+            self.apps = [  # In MSAL Python, each CCA binds to one tenant only
```

Comment:
> In MSAL Python, each CCA binds to one tenant only

Interesting. Do you mean 1 CCA per tenant or per authority? What about multi-tenant apps or org / consumer / common authorities?

In .NET tests we just use 1 CCA instance with default authority (common) since it doesn't matter for tests..